### PR TITLE
DCGEO-97: MCS claims RDA API caller and sink

### PR DESCRIPTION
**Pull request by brianburton**
**This PR is still a WIP but has been created to allow early review by DCGEO team members.**

### Change Details

This PR implements DCGEO-97 to add a new PipelineJob that calls the RDA API `getMcsClaims()` RPC and stores all claims in the resulting stream to the database.  The chane also implements the entity and protocol changes from the RDA API 0.3 MVP release.

Some of the existing code specific to FISS claim processing in master has been converted to apply to both FISS and MCS claims.  In particular:

- RDA API proto files have been updated to the 0.3 MVP release.
- A new schema migration script has been added to create the MCS tables and add new columns to the FISS tables.
- New entity classes have been added for MCS claims.  These are the same entities added for [PR 677](https://github.com/CMSgov/beneficiary-fhir-data/pull/677).
- FISS entities have been modified to include the new columns added by RDA API.
- `RdaChange` has been added to model the new RDA API `ClaimChange` wrapper object.
- `FissClaimRdaSink` has been renamed `JpaClaimRdaSink` and converted to accept the claim entity class as a type parameter.  Now this same sink is used for both FISS and MCS claims.
- The previous `PipelineJob` implementation, `RdaLoadJob` has been converted to an abstract class, `AbstractRdaLoadJob`, and trivial new subclasses have been defined for FISS and MCS (`RdaFissClaimLoadJob` and `RdaMcsClaimLoadJob`, respectively).  The subclassing was necessary to allow the `PipelineJobType` to be based on the job's class.
- A number of new integration tests have been added, including  `JpaClaimRdaSinkIT` which verifies claim entities can be written to HSQLDB correctly.
- The mock RDA API server has been extended to support `getMcsClaims` calls.

Key classes to review in this PR:

- `V31__preadj_claim_table_updates.sql` migration script.
- `McsClaimTransformer` validates and converts RDA API `McsClaim` objects into equivalent `PreAdjMcsClaim` entity objects.
- `FissClaimTransformer` validates and converts RDA API `FissClaim` objects into equivalent `PreAdjFissClaim` entity objects.  This class has been modified to add support for new fields.
- `DataTransformer` has been enhanced to handle new types in MCS claim data.
- `EnumStringExtractor` implements the special handling required for RDA API enum values.
- `JpaClaimRdaSink` writes JPA entities to database in batches.  Same as previous `FissClaimRdaSink` but now entity type is a generic type parameter on the class.  Now uses change type from `RdaChange` to optimize write perfomance.
- `PipelineApplication` has been modified to register both a `RdaMcsClaimLoadJob` and a `RdaFissClaimLoadJob`.

### Acceptance Validation

Nothing in particular.

### Feedback Requested

Nothing in particular.

### External References

- [DCGEO-97](https://jira.cms.gov/browse/DCGEO-97)

### Security Implications

The new pipeline job will download MCS claims from the RDA API server and store them in the database.

- [ ] new software dependencies

The `bfd-pipeline-rda-grpc` module has a new (for it) compile time dependency on lombok.  This is the same library used by `bfd-model-rda` so it's not new to BFD.

- [ ] altered security controls

None

- [ ] new data stored or transmitted

MCS claims from RDA API will now be stored to postgres when the RDA pipeline has been enabled.

- [ ] security checklist is completed for this change

No

- [ ] requires more information or team discussion to evaluate security implications

No

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java [hunk 1]
```diff
@@ -43,11 +43,12 @@
   /**
    * The number of jobs that can be run at one time. Because the {@link VolunteerJob} and {@link
    * SchedulerJob} will always be running, this number must be greater than or equal to 3, in order
-   * for any actual jobs to get run.
+   * for any actual jobs to get run. Value of 5 is sufficient for VolunteerJob, SchedulerJob,
+   * CcwRifLoadJob, RdaFissClaimLoadJob, and RdaMcsClaimLoadJob.
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:47]: thanks for listing these, thats helpful context

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java [hunk 2]
```diff
@@ -86,40 +72,55 @@ public void close() throws Exception {
...
-        entityManager.persist(claim);
+      for (RdaChange<TClaim> change : changes) {
+        switch (change.getType()) {
+          case INSERT:
+            entityManager.persist(change.getClaim());
+            break;
+          case UPDATE:
+            entityManager.merge(change.getClaim());
+            break;
+          case DELETE:
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java:120]:
```suggestion
          case DELETE:
            // TODO: [DCGEO-131] accept DELETE changes from RDA API
```

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java [hunk 3]
```diff
@@ -131,12 +132,16 @@ private void persistBatch(Iterable<PreAdjFissClaim> claims) {
...
+  private void mergeBatch(Iterable<RdaChange<TClaim>> changes) {
     boolean commit = false;
     try {
       entityManager.getTransaction().begin();
-      for (PreAdjFissClaim claim : claims) {
-        entityManager.merge(claim);
+      for (RdaChange<TClaim> change : changes) {
+        if (change.getType() != RdaChange.Type.DELETE) {
+          entityManager.merge(change.getClaim());
+        } else {
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java:142]:
```suggestion
        } else {
          // TODO: [DCGEO-131] accept DELETE changes from RDA API
```

**Review by ericbuckley — COMMENTED** [review]
looking great so far, thanks for being so diligent in mapping all of the new fields from the RDA API.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java [hunk 4]
```diff
@@ -18,25 +18,37 @@
...
+    final String claimType = args[1];
+
     final RdaLoadOptions jobConfig = readRdaLoadOptionsFromProperties(props);
     final DatabaseOptions databaseConfig = readDatabaseOptions(props);
     final PipelineApplicationState appState =
         new PipelineApplicationState(
             new MetricRegistry(), databaseConfig, RDA_PERSISTENCE_UNIT_NAME);
-    final PipelineJob job = jobConfig.createFissClaimsLoadJob(appState);
+    final PipelineJob<?> job =
+        claimType.equals("fiss")
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java:49]: is there another check to see if claimtype is "fiss" or "mcs" and if not raise an error?  If the intention of this app is ONLY for testing purposes than you can ignore, just want to make sure nothing happens in the future where "fiss" is mistyped (eg, "fis") and the mcs claims load runs

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/JsonClaimSource.java [hunk 5]
```diff
@@ -35,21 +38,48 @@ public JsonFissClaimSource(String json) {
...
+    FissClaim.Builder claim = FissClaim.newBuilder();
+    JsonFormat.parser().merge(jsonString, claim);
+    return claim.build();
+  }
+
+  /**
+   * This method fits the signature of Parser&lt;McsClaim&gt;
+   *
+   * @param jsonString JSON to be parsed
+   * @return a FissClaim object
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/JsonClaimSource.java:76]:
```suggestion
   * @return a McsClaim object
```

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    if (entityManager != null && entityManager.isOpen()) {
+      entityManager.close();
+    }
+  }
+
+  public int writeBatch(Collection<RdaChange<TClaim>> claims) throws ProcessingException {
+    try {
+      metrics.calls.mark();
+      try {
+        persistBatch(claims);
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java:45]: I'm thinking one day we might want to optimize this call differently for FISS and MCS records, I'm thinking thats achievable by using different batch sizes, in which case I think that's already built-in to your implementation, but wanted to double-check that each sink could be configured to use different batch sizes in the same env?

**Review by ericbuckley — DISMISSED** [review]
thanks for building the changes out to the mock server as well, also great job with the test cases

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java (hunk 4 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/DirectRdaLoadApp.java:49]: Good point.  It is just for testing but it's easy enough to catch this.  I'll fix.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java (hunk 6 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/JpaClaimRdaSink.java:45]: Good question.  Currently both pipelines share the same settings for batch size, run interval, etc.  Can we defer that change to a later sprint?  I think load testing will reveal whether it's necessary or not.

**Review by KyRobbins — APPROVED** [review]

**Review by ericbuckley — APPROVED** [review]
👍

**Review by cbrunefearless — APPROVED** [review]
lgtm

**Pull request merged**